### PR TITLE
Don't use le32/le64

### PR DESCRIPTION
**Pull request by steven-johnson**
le32/le64 is being removed from LLVM19, which is bad for Halide. This uses i386/x86-64 and wasm32/wasm64 targets instead, with a modification to clang flags. 

NOTE that if we go in this direction, it means that LLVM used for building Halide *must* have WebAssembly codegen enabled.

### CODE REVIEW

**Comment by steven-johnson** [comment]
Oddly, it seems to work here, but inside google3 it fails massively, with ~every runtime call reported as missing

**Comment by mcourteaux** [comment]
Ah, this is the fix for #8333 I guess?

File: src/runtime/profiler_common.cpp [hunk 1]
```diff
@@ -350,7 +349,11 @@ WEAK void halide_profiler_stack_peak_update(void *user_context,
     // Update per-func memory stats
     for (int i = 0; i < instance->pipeline_stats->num_funcs; ++i) {
         if (f_values[i] != 0) {
-            sync_compare_max_and_swap(&(instance->funcs[i]).stack_peak, f_values[i]);
+            // On 32-bit platforms we don't want to use 64-bit
+            // atomics. Fortunately on these platforms memory usage fits into
+            // 32-bit integers.
+            sync_compare_max_and_swap((uintptr_t *)(&(instance->funcs[i]).stack_peak),
```
Comment (steven-johnson) [inline src/runtime/profiler_common.cpp:355]: This isn't safe or correct. stack_peak is a uint64, so the result will be incorrect on 32-bit platforms.

Comment (abadams) [reply src/runtime/profiler_common.cpp]: This assumes little-endian. It will just update the low 32 bits of stack_peak. The high bits will be zero on 32-bit targets anyway.

**Comment by steven-johnson** [comment]
Mac-specific failures appear to be due to a transitory bug in top-of-tree Clang that was improperly coalescing literal strings (!)

File: src/runtime/CMakeLists.txt [hunk 2]
```diff
@@ -217,17 +217,31 @@ foreach (i IN LISTS RUNTIME_CPP)
                     # unfortunately, clang doesn't automatically set this flag even though the
                     # ABI is msvc on windows
                     set(fshort-wchar -fshort-wchar)
-                    set(TARGET "le64-unknown-windows-unknown")
+                    if (LLVM_PACKAGE_VERSION VERSION_LESS 19.0)
+                        set(TARGET "le64-unknown-windows-unknown")
+                    else ()
+                        # TODO: was le64 here, not sure if this is correct or not
+                        set(TARGET "aarch64-unknown-windows-unknown")
+                    endif ()
```
Comment (alexreinking) [inline src/runtime/CMakeLists.txt:225]: What prevents us from just using `aarch64`/`arm` on all LLVM versions? I'm a bit worried about bugs in Halide 19 that appear only with LLVM 19 and not LLVM 18.

Also, does this introduce a requirement that LLVM be built with `ARM` and `AArch64` backends? We will need to check that, if so. #2282 will rear its head again.

Comment (steven-johnson) [reply src/runtime/CMakeLists.txt:225]: > What prevents us from just using aarch64/arm on all LLVM versions?

Nothing AFAICT. I'll make the change.

> Also, does this introduce a requirement that LLVM be built with ARM and AArch64 backends

Yes.

File: src/runtime/profiler_common.cpp (hunk 1 above)
Comment (steven-johnson) [reply src/runtime/profiler_common.cpp]: Wouldn't it be cleaner to just change `stack_peak` to be a uintptr_t?

Comment (abadams) [reply src/runtime/profiler_common.cpp]: I was worried about breaking existing users of the profiler stats.

**Comment by steven-johnson** [comment]
This change appears to be working in both GitHub and google3, PTAL

File: Makefile [hunk 3]
```diff
@@ -1093,7 +1094,8 @@ RUNTIME_CXX_FLAGS = \
     -Wno-unknown-warning-option \
     -Wno-unused-function \
     -Wvla \
-    -Wsign-compare
+    -Wsign-compare \
+    -Wno-sync-alignment
```
Comment (abadams) [inline Makefile:1098]: Is this change still needed?

File: Makefile [hunk 4]
```diff
@@ -1068,7 +1068,8 @@ RUNTIME_TRIPLE_WIN_X86_32 = "i386-unknown-windows-unknown"
 RUNTIME_TRIPLE_WIN_X86_64 = "x86_64-unknown-windows-unknown"
 RUNTIME_TRIPLE_WIN_ARM_32 = "arm-unknown-windows-unknown"
 RUNTIME_TRIPLE_WIN_ARM_64 = "aarch64-unknown-windows-unknown"
-RUNTIME_TRIPLE_WIN_GENERIC_64 = "le64-unknown-windows-unknown"
+# TODO: was le64 here, not sure if this is correct or not
+RUNTIME_TRIPLE_WIN_GENERIC_64 = "aarch64-unknown-windows-unknown"
```
Comment (alexreinking) [inline Makefile:1072]: This is inconsistent with the CMake build, which uses `x86_64` instead of `aarch64`.

**Review by alexreinking — CHANGES_REQUESTED** [review]
I think this needs to be discussed at the dev meeting because of the impact it has on target-subsetted builds.

File: src/runtime/CMakeLists.txt [hunk 5]
```diff
@@ -217,17 +218,17 @@ foreach (i IN LISTS RUNTIME_CPP)
                     # unfortunately, clang doesn't automatically set this flag even though the
                     # ABI is msvc on windows
                     set(fshort-wchar -fshort-wchar)
```
Comment (alexreinking) [inline src/runtime/CMakeLists.txt:220]: The comments here say the ABI is `msvc`... is that accurate? What would happen if we used `x86_64-unknown-windows-msvc` as the target? Would we still need the `-fshort-wchar` flag in that case?

**Comment by steven-johnson** [comment]
> I think this needs to be discussed at the dev meeting because of the impact it has on target-subsetted builds.

Agreed

File: Makefile (hunk 4 above)
Comment (steven-johnson) [reply Makefile:1072]: Fixed

File: src/runtime/CMakeLists.txt (hunk 5 above)
Comment (steven-johnson) [reply src/runtime/CMakeLists.txt:220]: I have no idea, that's a question for @slomp

File: Makefile (hunk 3 above)
Comment (steven-johnson) [reply Makefile:1098]: No, removing

File: src/runtime/CMakeLists.txt (hunk 5 above)
Comment (slomp) [reply src/runtime/CMakeLists.txt:219]: Yup, we need to adhere to the MSVC ABI because D3D12 is shipped as a DLL, and we need to talk to it with the "standard" Windows x64 ABI rules.
A long time ago, when d3d12 back-end shipped, I ended up implementing llvm assembly stubs to patch calls to D3D12 DLL, but in this PR https://github.com/halide/Halide/pull/5217 we decided that changing the ABI in the triplet was the right thing to do.

> Would we still need the -fshort-wchar flag in that case?
No idea. The `-fshort-wchar` documentation is abysmal, and non-MSVC compiler support for Windows shenanigans is janky at best.

Comment (alexreinking) [reply src/runtime/CMakeLists.txt:219]: > and non-MSVC compiler support for Windows shenanigans is janky at best.

IIRC we removed MinGW support ages ago and no one has asked for it back c.f. #4601

Comment (alexreinking) [reply src/runtime/CMakeLists.txt:219]: I'm not sure when we started setting that flag, exactly, but `sizeof(wchar_t)` has been `2` when the OS target is Windows since Clang 3.5

https://gcc.godbolt.org/z/3rWP7r8qT

File: Makefile (hunk 3 above)
Comment (steven-johnson) [reply Makefile:1098]: Update: yes, it is needed now

**Review by alexreinking — APPROVED** [review]
LGTM. We have a few alternatives if this change becomes a problem, such as working with package maintainers to include these backends.

**Comment by steven-johnson** [comment]
OK, I think this combination finally works, going to land it
